### PR TITLE
Add ring test. Also cargo run --release --example ping doesn't work with LTO

### PR DESCRIPTION
**Pull request by ehiggs**
This should offer a benchmark that can be run against each version and satisfied much of #52.

### CODE REVIEW

**Comment by ehiggs** [comment]
I took off lto from profile.release as this breaks examples:

```
$ cargo run --release --example ping
   Compiling actix v0.7.11 (/Users/ehigg90120/src/download/actix)
error: can't perform LTO when compiling incrementally

error: aborting due to previous error

error: Could not compile `actix`.

To learn more, run the command again with --verbose.
```

**Comment by DoumanAsh** [comment]
That's weird, can you remove `profile.release` section completely and see that it works with lto?

opt-level and lto with disabled incremental compilation is anyway default

**Comment by exgens** [comment]
There was an experiment: https://github.com/rust-lang/cargo/pull/6564. Maybe this is the reason?
@ehiggs what version of Rust you using?

**Comment by DoumanAsh** [comment]
Oh damn, suddenly changing defaults... 

@exgens do you know since when it will be included? Is it going to be 1.34?

**Comment by ehiggs** [comment]
```
$ cargo --version
cargo 1.34.0-nightly (245818076 2019-01-27)
```

**Comment by exgens** [comment]
It was just an experiment for a week to collect the data. It's already disabled.
There is a PR for ignore incremental when lto is set https://github.com/rust-lang/cargo/pull/6643 so I think you don't need to worry about it 🙂
But some projects (for example https://github.com/jwilm/alacritty) already disabled incremental for release builds with lto enabled: https://github.com/jwilm/alacritty/commit/7eb0ea82ef491f277b8278b64272dd7f245c5953 🙂

**Comment by ehiggs** [comment]
@exgens 
>It was just an experiment for a week to collect the data. It's already disabled.

Should I remote the lto removal then?

**Comment by ehiggs** [comment]
I've upgraded nightly again and indeed the experiment is over. Sorry about that!

I've pushed another commit that puts LTO back in. I also added a comment explaining what the example does.

The performance is a bit all over the place, though.

```
$ cargo run --release --example ring 1000 1000
    Finished release [optimized] target(s) in 11.32s
     Running `target/release/examples/ring 1000 1000`
Setting up nodes
Reached limit of 1000000 (payload was 1000000)
Time taken: 2.129238
$ cargo run --release --example ring 1000 1000
    Finished release [optimized] target(s) in 0.40s
     Running `target/release/examples/ring 1000 1000`
Setting up nodes
Reached limit of 1000000 (payload was 1000000)
Time taken: 1.987859
$ cargo run --release --example ring 1000 1000
    Finished release [optimized] target(s) in 0.29s
     Running `target/release/examples/ring 1000 1000`
Setting up nodes
Reached limit of 1000000 (payload was 1000000)
Time taken: 2.729968
```

File: examples/ring.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            prev_addr = Node{limit: num_nodes * ntimes, next: prev_addr.recipient()}.start();
+        }
+
+        Node{limit: num_nodes * ntimes, next: prev_addr.recipient()}
+    });
+
+    let now = SystemTime::now();
+    system.run();
+    match now.elapsed() {
+        Ok(elapsed) => println!("Time taken: {}.{:06}", elapsed.as_secs(), elapsed.subsec_micros()),
```
Comment (DoumanAsh) [inline examples/ring.rs]: Could you please add units in this println

File: examples/ring.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        if msg.0 >= self.limit {
+            println!("Reached limit of {} (payload was {})", self.limit, msg.0);
+            System::current().stop();
+            return;
+        }
+        self.next.do_send(Payload(msg.0 + 1)).expect("Unable to send payload");
+    }
+}
+
+fn print_usage_and_exit() {
```
Comment (DoumanAsh) [inline examples/ring.rs]: `print_usage_and_exit() -> !`

File: examples/ring.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            ::std::process::exit(1);
+        }
+        num_nodes = arg_num_nodes;
+    } else {
+        print_usage_and_exit();
+    }
+    let num_nodes = num_nodes;
+
+    let mut ntimes = 1;
+    if let Ok(arg_ntimes) = args[2].parse::<usize>() {
```
Comment (DoumanAsh) [inline examples/ring.rs]: `let n_times = if let Ok(arg_`

File: examples/ring.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        ntimes = arg_ntimes;
+    } else {
+        print_usage_and_exit();
+    }
+    let ntimes = ntimes; // make constant;
+
+    let system = System::new("test");
+
+
```
Comment (DoumanAsh) [inline examples/ring.rs]: Remove extra empty line

File: examples/ring.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    println!("Setting up nodes");
+    let _: Addr<_> = Node::create(move|ctx| {
```
Comment (DoumanAsh) [inline examples/ring.rs]: is `Addr<_>` necessary?

Please remove all explicit types if they are not necessary here and below

**Comment by DoumanAsh** [comment]
Thanks, the thing with performance is quite funny.

I imagine multi-threaded runtime might bring a better result for such test, while with single threaded tokio runtime I'd expect more stable results...

**Pull request merged**